transforms/ember-object/index.js

```diff
@@ -0,0 +1,197 @@
+module.exports = function transformer(file, api) {
+  const j = api.jscodeshift;
+  const root = j(file.source);
+
+  // Helper functions
+  /**
+   * Transform instance properties to MemberExpressions
+   *
+   * For example: `prop: value` --> `this.prop = value`
+   *
+   * @param {*} instanceProps
+   * @returns Array [ ExpressionStatement ]
+   */
+  function instancePropsToExpressions(instanceProps) {
+    return instanceProps.map(instanceProp =>
+      j.expressionStatement(
+        j.assignmentExpression(
+          "=",
+          j.memberExpression(j.thisExpression(), instanceProp.key),
+          instanceProp.value
+        )
+      )
+    );
+  }
+
+  /**
+   * Transform functions to class methods
+   *
+   * For example { foo: function() { }} --> { foo() { }}
+   *
+   * @param {[]} functionProps
+   * @returns Array [MethodDefinition]
+   */
+  function functionPropsToMethods(functionProps = []) {
+    return functionProps.map(functionProp => {
+      return j.methodDefinition("method", functionProp.key, functionProp.value);
+    });
+  }
+
+  /**
+   * Create  a constructor method
+   *
+   * @param {[]} instanceProps
+   * @return Array [MethodDefinition]
+   */
+  function createConstructor(instanceProps = []) {
+    if (instanceProps.length) {
+      return [
+        j.methodDefinition(
+          "constructor",
+          j.identifier("constructor"),
+          j.functionExpression(
+            null,
+            [],
+            j.blockStatement(instancePropsToExpressions(instanceProps))
+          )
+        )
+      ]; // Todo: find a way to include params
+    }
+
+    return [];
+  }
+
+  /**
+   * Create the class
+   *
+   * @param {String} className
+   * @param {Object} { instanceProps, functionProps } map of instance and function properties
+   */
+  function createClass(className, { instanceProps, functionProps } = {}) {
+    return j.classDeclaration(
+      j.identifier(className),
+      j.classBody(
+        []
+          .concat(createConstructor(instanceProps))
+          .concat(functionPropsToMethods(functionProps))
+      ),
+      j.identifier("EmberObject")
+    );
+  }
+
+  /**
+   * Return the map of instance props and functions from Ember Object
+   *
+   * For example
+   * const myObj = EmberObject.extend({ key: value, foo: function() {}, baz() {} });
+   * will be parsed as:
+   *{
+      instanceProps: [ key: value ]
+      functionProps: [ FunctionExpression(foo), FunctionExpression(baz) ]
+    }
+   * @param {*} emberObjectExpression
+   * @returns { instanceProps, functionProps } map of instance and function properties
+   */
+  function getEmberObjectProperties(emberObjectExpression) {
+    const objProps = j(emberObjectExpression)
+      .find(j.ObjectExpression)
+      .get("properties").value;
+
+    const instanceProps = [];
+    const functionProps = [];
+
+    objProps.forEach(prop => {
+      if (prop.value.type === "FunctionExpression") {
```

Comment:
Although unlikely, we _may_ have to deal with native accessors as well (`get foo`, `set foo`).

Reply:
Thanks for the comment! So far I haven't encountered them in the code which I have seen, but will add a logic for it. Added a TODO as of now.

Reply:
I changed the fixtures to use exactly mentioned in @pzuraq 's issue #1 I have a question though

```
const Foo = EmberObject.extend({
  prop: "defaultValue",
});
```
Should be transformed to 
```
class Foo extends EmberObject {
  prop = 'defaultValue';
}
```
OR
```
class Foo extends EmberObject {
  constructor() {
    super();
    this.prop = "defaultValue";
  }
}
```
@pzuraq @scalvert

Reply:
I think this depends on whether or not we want to have a transform level with only stage 4+ JS features. If so, we cannot use class fields for that level.

Personally, class fields feel much nicer (and will be much easier to codemod to/from in the future). Given that they're stage 3, and that public fields especially are very well understood, I'd say we should start by assuming that at the least users will be ok with stage 3+, and add support for stage 4+ later on if there is demand for it.

Reply:
@pzuraq  I am not sure I understand what you mean by stage 3 vs stage 4 features. Sorry about that!

Does it mean we have to go with
```
class Foo extends EmberObject {
  constructor() {
    super();
    this.prop = "defaultValue";
  }
}
```

Reply:
Ah for sure, no worries! The stages are referring to the [TC39 process](https://tc39.github.io/process-document/), which is the process that is used to add new features to JS. You can see the current status of the proposals [here](https://github.com/tc39/proposals).

Stage 4 means completed, stage 3 means mostly completed, and stage 2 means it will most likely make it into the language, but could still change pretty significantly. Currently, class fields are at stage 3, and decorators are at stage 2.

So, the question here is do we create 3 levels of the transform (stage 4+, stage 3+, stage 2+) or do we create 2 levels (stage 3+, stage 2+). Creating 3 levels is extra work, so I would lean toward starting with just 2 levels and using the class fields style:

```js
class Foo extends EmberObject {
  prop = 'defaultValue';
}
```

And adding the stage 4+ level if we decide it's necessary later on. What do you think @scalvert?